File: backend/app/schemas/personale.py
import uuid
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, model_validator

from app.schemas.registro_campi import VerificationStatus


class _OrmModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class AnaPersoneCreate(_OrmModel):
    fotografia: str | None = None
    cognome: str
    nome: str
    sesso: str | None = None
    data_nascita: date | None = None
    luogo_nascita: str | None = None
    nazionalita: str | None = None
    conoscenza_lingua_italiana: str | None = None
    codice_fiscale: str
    residenza: str | None = None
    tipologia_contratto: str | None = None
    data_assunzione: date | None = None
    data_fine_rapporto: date | None = None
    mansione: str | None = None
    persona_backup_id: uuid.UUID | None = None
    processi_speciali_eseguiti: str | None = None
    conoscenza_organizzazione_livello_id: uuid.UUID | None = None
    competenze_livello_id: uuid.UUID | None = None
    consapevolezza_livello_id: uuid.UUID | None = None
    frequenza_visite_mediche: int | None = None
    altro: str | None = None
    note: str | None = None


class AnaPersoneUpdate(_OrmModel):
    fotografia: str | None = None
    cognome: str | None = None
    nome: str | None = None
    sesso: str | None = None
    data_nascita: date | None = None
    luogo_nascita: str | None = None
    nazionalita: str | None = None
    conoscenza_lingua_italiana: str | None = None
    codice_fiscale: str | None = None
    residenza: str | None = None
    tipologia_contratto: str | None = None
    data_assunzione: date | None = None
    data_fine_rapporto: date | None = None
    mansione: str | None = None
    persona_backup_id: uuid.UUID | None = None
    processi_speciali_eseguiti: str | None = None
    conoscenza_organizzazione_livello_id: uuid.UUID | None = None
    competenze_livello_id: uuid.UUID | None = None
    consapevolezza_livello_id: uuid.UUID | None = None
    frequenza_visite_mediche: int | None = None
    altro: str | None = None
    note: str | None = None


class AnaPersoneRead(AnaPersoneCreate):
    id: uuid.UUID
    azienda_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class AnaPersoneGiuridicheCreate(_OrmModel):
    denominazione: str
    codice_fiscale: str
    partita_iva: str | None = None
    sede: str | None = None
    note: str | None = None


class AnaPersoneGiuridicheUpdate(_OrmModel):
    denominazione: str | None = None
    codice_fiscale: str | None = None
    partita_iva: str | None = None
    sede: str | None = None
    note: str | None = None


class AnaPersoneGiuridicheRead(AnaPersoneGiuridicheCreate):
    id: uuid.UUID
    azienda_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class PersonaSummary(_OrmModel):
    """Dati anagrafici per la compilazione automatica e sola lettura delle
    righe incarico (specifica CCIAA §5.2/5.3/6.2: nome, nascita, cittadinanza
    e domicilio del soggetto sono "letti dal soggetto", mai duplicati o
    modificabili nel record della carica). Nessuna colonna nuova: sono già
    tutte presenti su `ana_persone`, prima non esposte qui."""

    id: uuid.UUID
    nome: str
    cognome: str
    codice_fiscale: str
    data_nascita: date | None = None
    luogo_nascita: str | None = None
    nazionalita: str | None = None
    residenza: str | None = None


class PersonaGiuridicaSummary(_OrmModel):
    """Come `PersonaSummary` sopra, per un titolare persona giuridica (§
    Correzione 16) — nessun campo di persona fisica (nascita, cittadinanza,
    domicilio): solo i dati identificativi del soggetto giuridico."""

    id: uuid.UUID
    denominazione: str
    codice_fiscale: str
    partita_iva: str | None = None
    sede: str | None = None


class RuoloSummary(_OrmModel):
    id: uuid.UUID
    codice: str
    codice_documento: str | None = None
    denominazione: str


class CaratteristicaRuoloRead(_OrmModel):
    """Configurazione di una caratteristica per un ruolo (join
    rel_ruoli_caratteristiche + cat_caratteristiche_incarico): usata dal
    frontend per costruire il form dinamico di un incarico senza duplicare
    lato client le regole già lette da `app.core.incarichi`."""

    id: uuid.UUID
    codice: str
    denominazione: str
    tipoDato: str
    valoriAmmessi: list[str] | None = None
    obbligatorieta: str


class IncaricoCreate(_OrmModel):
    # § Correzione 16: esattamente uno dei due titolari, mai entrambi né
    # nessuno — stesso vincolo di `chk_per_incarichi_titolare_esclusivo` a
    # livello DB, verificato qui per un errore di validazione immediato
    # invece di un errore di integrità del database.
    persona_id: uuid.UUID | None = None
    persona_giuridica_id: uuid.UUID | None = None
    ruolo_id: uuid.UUID
    note: str | None = None
    valori: dict[str, Any] = {}

    @model_validator(mode="after")
    def _titolare_esclusivo(self) -> "IncaricoCreate":
        if (self.persona_id is None) == (self.persona_giuridica_id is None):
            raise ValueError("Specificare esattamente uno tra persona_id e persona_giuridica_id.")
        return self


class IncaricoUpdate(_OrmModel):
    persona_id: uuid.UUID | None = None
    persona_giuridica_id: uuid.UUID | None = None
    ruolo_id: uuid.UUID | None = None
    note: str | None = None
    valori: dict[str, Any] | None = None

    @model_validator(mode="after")
    def _titolare_non_ambiguo(self) -> "IncaricoUpdate":
        if self.persona_id is not None and self.persona_giuridica_id is not None:
            raise ValueError("persona_id e persona_giuridica_id non possono essere valorizzati insieme.")
        return self


class IncaricoRead(_OrmModel):
    id: uuid.UUID
    azienda_id: uuid.UUID
    persona_id: uuid.UUID | None = None
    persona_giuridica_id: uuid.UUID | None = None
    ruolo_id: uuid.UUID
    note: str | None = None
    valori: dict[str, Any] = {}
    persona: PersonaSummary | None = None
    persona_giuridica: PersonaGiuridicaSummary | None = None
    ruolo: RuoloSummary
    created_at: datetime
    updated_at: datetime
    # Verifica del consulente sulla riga (vedi app/core/incarichi.py): non
    # più la caratteristica A32 dentro il form, ma lo stesso trattamento del
    # registro campo-per-campo (nota, audit, concorrenza ottimistica).
    verificationStatus: VerificationStatus | None = None
    verificationVersion: int | None = None
    revisionNote: str | None = None
    verifiedAt: str | None = None
    verifiedBy: str | None = None

File: frontend/lib/types/personale.ts
// Tipi del motore generico "incarico" (persona + ruolo + caratteristiche,
// vedi backend/app/schemas/personale.py e backend/app/core/incarichi.py):
// sostituisce le tabelle qual_* per le card Soci/Amministratori/Sindaci
// della griglia CCIAA. Solo persone fisiche fino alla Correzione 16
// (decisione utente 2026-08-26); da quella correzione un incarico può
// avere anche un titolare persona GIURIDICA (`AnaPersonaGiuridica`) — vedi
// `persona_giuridica_id`/`persona_giuridica` su `Incarico` sotto.

import type { VerificationStatus } from "@/lib/types/registro";

export type AnaPersona = {
  id: string;
  azienda_id: string;
  cognome: string;
  nome: string;
  codice_fiscale: string;
  data_nascita: string | null;
  sesso: string | null;
  luogo_nascita: string | null;
  nazionalita: string | null;
  created_at: string;
  updated_at: string;
};

export type PersonaCreatePayload = {
  cognome: string;
  nome: string;
  codice_fiscale: string;
  data_nascita?: string | null;
  luogo_nascita?: string | null;
  nazionalita?: string | null;
  residenza?: string | null;
};

export type PersonaSummary = {
  id: string;
  nome: string;
  cognome: string;
  codice_fiscale: string;
  data_nascita: string | null;
  luogo_nascita: string | null;
  nazionalita: string | null;
  residenza: string | null;
};

// § Correzione 16: persona giuridica (società, ente) — titolare alternativo
// di un incarico quando il ruolo può essere affidato a un soggetto esterno
// anziché a una persona fisica (primo caso: "Società di revisione legale").
// Nessun campo di persona fisica (nascita, cittadinanza): solo i dati
// identificativi del soggetto giuridico.
export type AnaPersonaGiuridica = {
  id: string;
  azienda_id: string;
  denominazione: string;
  codice_fiscale: string;
  partita_iva: string | null;
  sede: string | null;
  note: string | null;
  created_at: string;
  updated_at: string;
};

export type PersonaGiuridicaCreatePayload = {
  denominazione: string;
  codice_fiscale: string;
  partita_iva?: string | null;
  sede?: string | null;
  note?: string | null;
};

export type PersonaGiuridicaSummary = {
  id: string;
  denominazione: string;
  codice_fiscale: string;
  partita_iva: string | null;
  sede: string | null;
};

export type RuoloSummary = {
  id: string;
  codice: string;
  codice_documento: string | null;
  denominazione: string;
};

// Corrisponde a cat_caratteristiche_incarico.tipo_dato: guida il controllo
// da usare nel form dinamico (§2.5 della specifica CCIAA).
export type TipoDatoCaratteristica =
  | "TESTO"
  | "TESTO_LUNGO"
  | "CATALOGO"
  | "CATALOGO_MULTIPLO"
  | "NUMERO"
  | "DATA"
  | "BOOLEANO"
  | "DOCUMENTO";

export type CaratteristicaRuolo = {
  id: string;
  codice: string;
  denominazione: string;
  tipoDato: TipoDatoCaratteristica;
  valoriAmmessi: string[] | null;
  obbligatorieta: "OBBLIGATORIA" | "CONDIZIONALE" | "FACOLTATIVA";
};

// I valori di un incarico transitano come stringa (testo/numero/data) o
// booleano/array, secondo il tipo_dato — stesso principio di "un solo tipo
// verificato dal backend" già in uso nel registro campo-per-campo.
export type ValoreIncarico = string | boolean | string[] | null;

export type Incarico = {
  id: string;
  azienda_id: string;
  // § Correzione 16: esattamente uno dei due è valorizzato (mai entrambi,
  // mai nessuno) — stesso vincolo di chk_per_incarichi_titolare_esclusivo
  // lato backend.
  persona_id: string | null;
  persona_giuridica_id: string | null;
  ruolo_id: string;
  note: string | null;
  valori: Record<string, ValoreIncarico>;
  persona: PersonaSummary | null;
  persona_giuridica: PersonaGiuridicaSummary | null;
  ruolo: RuoloSummary;
  created_at: string;
  updated_at: string;
  // Verifica del consulente sulla riga (vedi backend/app/core/incarichi.py):
  // non più la caratteristica A32 dentro il form, stesso trattamento del
  // registro campo-per-campo (nota, audit, concorrenza ottimistica).
  verificationStatus: VerificationStatus | null;
  verificationVersion: number | null;
  revisionNote: string | null;
  verifiedAt: string | null;
  verifiedBy: string | null;
};

export type IncaricoPayload = {
  // § Correzione 16: esattamente uno dei due, mai entrambi — verificato sia
  // qui (chi costruisce il payload) sia lato backend (IncaricoCreate).
  persona_id?: string | null;
  persona_giuridica_id?: string | null;
  ruolo_id: string;
  note?: string | null;
  valori: Record<string, ValoreIncarico>;
};
